### PR TITLE
Make query result push rows to the server response

### PR DESCRIPTION
**Pull request by thobe**

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/graphdb/Result.java [hunk 1]
```diff
@@ -227,15 +227,17 @@
     /**
      * This is the visitor interface you need to implement to use the {@link Result#accept(ResultVisitor)} method.
      */
-    interface ResultVisitor
+    interface ResultVisitor<VisitationException extends Exception>
```
Comment (pontusmelke) [inline community/kernel/src/main/java/org/neo4j/graphdb/Result.java:232]: Could we have something like 

```
interface ResultsVisitor<E> extends org.neo4j.helpers.collection.Visitor<Result.ResultRow, E>
```

instead?

File: community/kernel/src/main/java/org/neo4j/graphdb/Result.java [hunk 2]
```diff
@@ -227,15 +227,17 @@
...
     {
         /**
          * Visits the specified row.
+         *
          * @param row the row to visit. The row object is only guaranteed to be stable until flow of control has
          *            returned from this method.
          * @return true if the next row should also be visited. Returning false will terminate the iteration of
          * result rows.
+         * @throws VisitationException
```
Comment (systay) [inline community/kernel/src/main/java/org/neo4j/graphdb/Result.java]: What is the contract when a VisitationException is thrown?

File: community/kernel/src/main/java/org/neo4j/graphdb/Result.java (hunk 1 above)
Comment (thobe) [inline community/kernel/src/main/java/org/neo4j/graphdb/Result.java:232]: I'd prefer not. It would be nice if we could treat the `org.neo4j.helpers` package as internal and not expose it in the API.

File: community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/ExpandC.scala [hunk 3]
```diff
@@ -22,35 +22,21 @@ package org.neo4j.cypher.internal.compiler.v2_3.birk.il
 import org.neo4j.graphdb.Direction
 
 case class ExpandC(fromVar: String, relVar: String, dir: Direction, types: Map[String, String], toVar: String, inner: Instruction) extends LoopDataGenerator {
-  private val visitMethod =
+  private val theBody =
```
Comment (systay) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/ExpandC.scala:25]: I really the changes in this file, but I believe it might not really belong in this commit, right?

Comment (thobe) [inline community/cypher/cypher-compiler-2.3/src/main/scala/org/neo4j/cypher/internal/compiler/v2_3/birk/il/ExpandC.scala:25]: It was required to make the tests pass...

I thought of putting in a separate commit in the same PR, but decided against it since it would have made the first commit useless on its own.
